### PR TITLE
feat: use RocksDB for state execution

### PR DESCRIPTION
**Pull request by KolbyML**
### What was wrong?
We are storing the whole Ethereum state in memory
### How was it fixed?

Store it in RocksDB.

We use sqlite for the rest of Trin, but this design choice doesn't make sense for executing the state. We need to use a KV store regardless it doesn't take more work to do it now, so I am doing it now.

Performance does matter and sqldb can't compete, executing the state is very strict when it comes to performance, so it just didn't make sense to use sql

### CODE REVIEW

File: portal-bridge/src/types/state/storage/evm_db.rs [hunk 1]
```diff
@@ -0,0 +1,245 @@
...
+    fn commit(&mut self, changes: HashMap<Address, Account>) {
+        for (address, mut account) in changes {
+            if !account.is_touched() {
+                continue;
+            }
+            if account.is_selfdestructed() {
+                let mut rocks_account: RocksAccount = match self
+                    .db
+                    .get(keccak256(address))
+                    .expect("Committing account to database should never fail")
```
Comment (morph-dev) [inline portal-bridge/src/types/state/storage/evm_db.rs:76]: nit: weird error message. Shouldn't it be `Reading account from...`?

File: portal-bridge/src/types/state/storage/evm_db.rs [hunk 2]
```diff
@@ -0,0 +1,245 @@
...
+        }
+        if account.code_hash == B256::ZERO {
+            account.code_hash = KECCAK_EMPTY;
+        }
+    }
+}
+
+impl DatabaseCommit for EvmDB {
+    fn commit(&mut self, changes: HashMap<Address, Account>) {
+        for (address, mut account) in changes {
```
Comment (morph-dev) [inline portal-bridge/src/types/state/storage/evm_db.rs:68]: nit: The `keccak256(address)` is used many times below. Many calculate it once at the start of the loop

**Review by morph-dev — APPROVED** [review]

File: portal-bridge/src/types/state/storage/account_db.rs [hunk 3]
```diff
@@ -0,0 +1,105 @@
...
+            address_hash: keccak256(address),
+            db,
+        }
+    }
+}
+
+fn combine_key(addr_hash: &[u8], key: &[u8]) -> Vec<u8> {
+    let mut dst = key.to_owned().to_vec();
+    {
+        for (k, a) in dst[12..].iter_mut().zip(&addr_hash[12..]) {
```
[the remark this answers is not in the document]
Comment (KolbyML) [reply portal-bridge/src/types/state/storage/account_db.rs:30]: I updated it to just concat like you said above

File: portal-bridge/src/types/state/execution.rs [hunk 4]
```diff
@@ -74,19 +73,20 @@ impl State {
...
+            let mut account = Account::default();
+            account.balance += alloc_balance.balance;
             self.database.trie.lock().insert(
                 keccak256(address).as_ref(),
                 &alloy_rlp::encode(AccountStateInfo {
-                    nonce: account_info.nonce,
-                    balance: account_info.balance,
+                    nonce: account.nonce,
+                    balance: account.balance,
                     storage_root: keccak256([EMPTY_STRING_CODE]),
```
[the remark this answers is not in the document]
Comment (KolbyML) [reply portal-bridge/src/types/state/execution.rs:83]: yeah it is only in the one place so I will leave it for now

File: portal-bridge/src/bridge/state.rs [hunk 5]
```diff
@@ -93,7 +94,7 @@ impl StateBridge {
         let gossip_send_semaphore = Arc::new(Semaphore::new(self.gossip_limit));
         let mut current_epoch_index = u64::MAX;
         let mut current_raw_era1 = vec![];
-        let mut state = State::new();
+        let mut state = State::new(Some(setup_temp_dir()?.into_path()));
```
[the remark this answers is not in the document]
Comment (KolbyML) [reply portal-bridge/src/bridge/state.rs:97]: good catch!

File: portal-bridge/src/types/state/storage/account_db.rs [hunk 6]
```diff
@@ -0,0 +1,99 @@
...
+
+impl AccountDB {
+    pub fn new(address: Address, db: Arc<RocksDB>) -> Self {
+        Self {
+            address_hash: keccak256(address),
+            db,
+        }
+    }
+
+    fn combine_key(addr_hash: &[u8], key: &[u8]) -> Vec<u8> {
```
Comment (morph-dev) [inline portal-bridge/src/types/state/storage/account_db.rs:26]: You could have made this:

```rust
fn get_db_key(&self, key: &[u8]) -> Vec<u8> {
  [self.address_hash.as_slice(), key].concat()
}
```

Then later, you would do: `self.db.get(self.get_db_key(key))`, `self.db.put(self.get_db_key(key), ...)` and `self.db.delete(self.get_db_key(key))`

File: portal-bridge/src/types/state/execution.rs (hunk 4 above)
[the remark this answers is not in the document]
Comment (morph-dev) [reply portal-bridge/src/types/state/execution.rs:83]: I counted at least 3 places :), but ok

File: portal-bridge/src/types/state/storage/account_db.rs (hunk 6 above)
Comment (KolbyML) [reply portal-bridge/src/types/state/storage/account_db.rs:26]: I will address this in my dao pr as well

**Pull request merged**